culane_metric/evaluate.py

```diff
@@ -0,0 +1,205 @@
+from scipy.optimize import linear_sum_assignment
+from shapely.geometry import LineString, Polygon
+
+import unsupervised_llamas.common.helper_scripts as helper_scripts
+import unsupervised_llamas.label_scripts.spline_creator as spline_creator
+
+LLAMAS_IMG_RES = (717, 1276)
+
+
+def add_ys(xs, vertical_cutoff=300):
+    assert len(xs) >= 717 - vertical_cutoff, "Need at least 417 pixels per lane"
+
+    # Reference lanes go from 0 to 717. If a horizontal entry is not
+    # defined, it is stored as -1. We have to filter for that.
+
+    if len(xs) == 717:  # lane regressed across complete image
+        xs = xs[vertical_cutoff:]
+    elif len(xs) == 417:  # lane regress across part of image that is relevant
+        pass
+    else:
+        raise NotImplementedError(f"Evaluations not implemented for length of detected lane: {len(xs)}")
+    xs = np.array(xs)
+    valid = xs >= 0
+    xs = xs[valid]
+    assert len(xs) > 1
+    ys = np.arange(300, 717)[valid]
+    return list(zip(xs, ys))
+
+
+def draw_lane(lane, img=None, img_shape=None, width=30):
+    if img is None:
+        img = np.zeros(img_shape, dtype=np.uint8)
+    lane = lane.astype(np.int32)
+    for p1, p2 in zip(lane[:-1], lane[1:]):
+        cv2.line(img, tuple(p1), tuple(p2), color=(1,), thickness=width)
+    return img
+
+
+def discrete_cross_iou(xs, ys, width=30, img_shape=LLAMAS_IMG_RES):
+    xs = [draw_lane(lane, img_shape=img_shape, width=width) > 0 for lane in xs]
+    ys = [draw_lane(lane, img_shape=img_shape, width=width) > 0 for lane in ys]
+
+    ious = np.zeros((len(xs), len(ys)))
+    for i, x in enumerate(xs):
+        for j, y in enumerate(ys):
+            ious[i, j] = (x & y).sum() / (x | y).sum()
+    return ious
+
+
+def continuous_cross_iou(xs, ys, width=30, img_shape=LLAMAS_IMG_RES):
+    h, w, _ = img_shape
+    image = Polygon([(0, 0), (0, h - 1), (w - 1, h - 1), (w - 1, 0)])
+    xs = [LineString(lane).buffer(distance=width / 2., cap_style=1, join_style=2).intersection(image) for lane in xs]
+    ys = [LineString(lane).buffer(distance=width / 2., cap_style=1, join_style=2).intersection(image) for lane in ys]
+
+    ious = np.zeros((len(xs), len(ys)))
+    for i, x in enumerate(xs):
+        for j, y in enumerate(ys):
+            ious[i, j] = x.intersection(y).area / x.union(y).area
+
+    return ious
+
+
+def interp(points, n=50):
+    x = [x for x, _ in points]
+    y = [y for _, y in points]
+    tck, _ = splprep([x, y], s=0, t=n, k=min(3, len(points) - 1))
+
+    u = np.linspace(0., 1., n)
+    return np.array(splev(u, tck)).T
+
+
+def culane_metric(pred, anno, width=30, iou_threshold=0.5, unofficial=False, img_shape=LLAMAS_IMG_RES):
+    if len(pred) == 0:
+        return 0, 0, len(anno)
+    if len(anno) == 0:
+        return 0, len(pred), 0
+    interp_pred = np.array([interp(pred_lane, n=50) for pred_lane in pred])  # (4, 50, 2)
+    interp_anno = np.array([interp(anno_lane, n=50) for anno_lane in anno])  # (4, 50, 2)
```

Comment:
What is interpolated here? The annotations already are interpolated splines, represented on a per pixel basis

Reply:
Indeed, the interpolation is not necessary for the annotations, only in the CULane dataset it is necessary.